### PR TITLE
[WIP]Cleanup scratch register

### PR DESCRIPTION
**Pull request by luyahan**
#412  
clean up scratch register and fixed t6 as rs1 of jalr or jal

### CODE REVIEW

**Comment by fw-brice** [comment]
Can you explain why t6 is removed from the set of scratch registers if it is still needed in several places for a scratch register? Doesn't that make the management of it more difficult?

**Comment by luyahan** [comment]
> Can you explain why t6 is removed from the set of scratch registers if it is still needed in several places for a scratch register? Doesn't that make the management of it more difficult?

@qjivy wang to fixed t6 as the rs1 of jalr.So we can know where jump from when occurs segment error.

**Comment by qjivy** [comment]
> Can you explain why t6 is removed from the set of scratch registers if it is still needed in several places for a scratch register? Doesn't that make the management of it more difficult?

One observation is that MIPS only uses t9 for jalr instruction (call function or jump to a code object). With this usage convention, if we have a failure site in the JITed code, we can simply regards the contents of t9 at this spot as the final jalr destination. This would be help debugging in a sense.

According to the above notion, we want to change all the jalr's rs register to t6. #420 tries to remove t6 from the MacroAssembler's scratch register list and use it for this special purpose. Avoiding to use t6 as a scratch register would help to reserve its value (not to be clobbered by instructions after the final jalr). It's a tiny improve for debugging. 

Please let us know if you have any ideas about this change. Perhaps it's better to firstly make a proposal next time.

**Comment by fw-brice** [comment]
Please check the precheck and resolve formatting issues to avoid problems upstreaming. I'll do it this time.

File: src/codegen/riscv64/assembler-riscv64.cc [hunk 1]
```diff
@@ -2739,8 +2737,8 @@ void Assembler::CheckTrampolinePool() {
         DCHECK(is_int32(imm64));
         int32_t Hi20 = (((int32_t)imm64 + 0x800) >> 12);
         int32_t Lo12 = (int32_t)imm64 << 20 >> 20;
-        auipc(scratch, Hi20);  // Read PC + Hi20 into t5.
-        jr(scratch, Lo12);     // jump PC + Hi20 + Lo12
+        auipc(t6, Hi20);  // Read PC + Hi20 into t5.
```
Comment (fw-brice) [inline src/codegen/riscv64/assembler-riscv64.cc:2740]: The comment should say "t6".

File: src/codegen/riscv64/macro-assembler-riscv64.cc [hunk 2]
```diff
@@ -1832,13 +1842,17 @@ void TurboAssembler::RoundHelper(FPURegister dst, FPURegister src,
...
 
-  // extract exponent value of the source floating-point to t6
-  if (std::is_same<F, double>::value) {
-    fmv_x_d(scratch, src);
-  } else {
-    fmv_x_w(scratch, src);
+  {
+    UseScratchRegisterScope temps2(this);
+    Register scratch = temps2.Acquire();
+    // extract exponent value of the source floating-point to t6
```
Comment (fw-brice) [inline src/codegen/riscv64/macro-assembler-riscv64.cc:1848]: Comment should say "scratch" instead of "t6".

**Review by fw-brice — CHANGES_REQUESTED** [review]
Just two changes to the comments to match the changed registers.

**Comment by fw-brice** [comment]
Merged

**Pull request closed without merging**